### PR TITLE
feat: support encrypted init segments

### PR DESCRIPTION
**Pull request by brandonocasey**
Support encrypted init segments

### CODE REVIEW

File: src/media-segment-request.js [hunk 1]
```diff
@@ -98,6 +98,15 @@ const handleErrors = (error, request) => {
     };
   }
 
+  if (request.responseType === 'arraybuffer' && request.response.byteLength === 0) {
```
Comment (brandonocasey) [inline src/media-segment-request.js:101]: Added this to `handleErrors` as it was something that was copy pasted for most of our functions.

File: test/media-segment-request.test.js [hunk 2]
```diff
@@ -708,6 +708,549 @@ QUnit.test('key 500 calls back with error', function(assert) {
...
+          }
+        }
+      }
+    },
+    trackInfoFn: this.noop,
+    timingInfoFn: this.noop,
+    dataFn: this.noop,
+    progressFn: this.noop,
+    doneFn: (error, segmentData) => {
+      assert.ok(error, 'error for invalid init segment');
```
Comment (gesinger) [inline test/media-segment-request.test.js:1062]: Where does the error happen here?

File: test/media-segment-request.test.js [hunk 3]
```diff
@@ -708,6 +708,549 @@ QUnit.test('key 500 calls back with error', function(assert) {
...
+      assert.ok(segmentData.bytes, 'decrypted bytes in segment');
+      assert.ok(segmentData.key.bytes, 'key bytes in segment');
+      assert.equal(
+        segmentData.key.bytes.buffer.byteLength,
+        16,
+        'key bytes are readable'
+      );
+
+      // verify stats
+      assert.equal(segmentData.stats.bytesReceived, 9, '6198 bytes');
```
Comment (gesinger) [inline test/media-segment-request.test.js:1228]:
```suggestion
      assert.equal(segmentData.stats.bytesReceived, 9, '9 bytes');
```

File: test/playback.test.js [hunk 4]
```diff
@@ -50,7 +50,7 @@ QUnit.module('Playback', {
 
     // uncomment these lines when deugging
     // videojs.log.level('debug');
-    this.fixture.style.position = 'inherit';
+    // this.fixture.style.position = 'inherit';
```
[the remark this answers is not in the document]
Comment (brandonocasey) [reply test/playback.test.js:53]: yeah this is currently how tests are running in main, we don't want to see players during the playback tests normally.

File: src/media-segment-request.js [hunk 5]
```diff
@@ -718,22 +756,32 @@ const waitForCompletion = ({
...
           doneFn
         });
+      };
+
+      // Keep track of when *all* of the requests have completed
+      segment.endOfAllRequests = Date.now();
+      if (segment.map && segment.map.encryptedBytes && !segment.map.bytes) {
+        return decrypt({
+          decryptionWorker,
+          id: segment.requestId,
```
[the remark this answers is not in the document]
Comment (brandonocasey) [reply src/media-segment-request.js:765]: might as well.

File: test/media-segment-request.test.js (hunk 2 above)
Comment (brandonocasey) [reply test/media-segment-request.test.js:1062]: it happens due to the decrypted init segment being webm.

Comment (gesinger) [reply test/media-segment-request.test.js:1062]: Makes sense 👍 . Might be worth a comment above it to point it out.

**Review by gesinger — APPROVED** [review]

**Pull request merged**